### DISCUSSION TITLE
Language server warnings with disabled automatic installation

### DISCUSSION THREAD

**Discussion by zenofile** [LunarVim/LunarVim#2930] [Q&A] [answered]
On the rolling branch, with `lvim.lsp.installer.setup.automatic_installation = false` on opening file types having an associated language server, neovim (NVIM v0.8.0-dev+905-g6b9852cc4) always shows the following warning two times:

    Spawning language server with cmd: `gopls` failed. The language server is either not installed, missing from PATH, or not executable.

![image](https://user-images.githubusercontent.com/2056412/185801270-79ce87b3-9529-4ae4-a59a-c2cd4ba592d4.png)

The same applies for `bashls` and others.

The only way I found this doesn't occur is explicitly disallowing the language servers via something like `vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "gopls", "bashls" })` even though they are not installed. Is this desired behavior or is there maybe a configuration error somewhere on my end?

**Comment by kylo252** [comment]
I'm not sure what the intended workflow here:

- do you want to enable `gopls`?
- do you want `gopls` to install automatically? if not, did you install `gopls`?

maybe it wasn't clear that server _installation_ is separate from server _configuration_

**Reply by zenofile** [reply]
My intention was to trigger the installation of any language server I actually need, manually, hence the disabled automatic installation.

But in doing so, there always seems to be an attempt from nvim to launch the corresponding language server on opening files even if I didn't manually install the server for that language and therefore doesn't exist. This leads to the display of the aforementioned warnings I suppose.

`gopls` was just intended as an example in the screenshot.

So in short, I want:

* trigger installation of language servers via `:LspInstall` manually only
* try not to launch a non-installed server on opening files (or not show a warning that it doesn't exist)

This “workflow” seemed to work until https://github.com/LunarVim/LunarVim/commit/eefde00ae80f91ecf88a93e869e346fdd04c7ba4 if I remember correctly.

**Comment by zenofile** [comment, marked as answer]
I got the desired behavior by introducing the following check in lua/lvim/lsp/manager.lua:

```patch
--- a/lua/lvim/lsp/manager.lua
+++ b/lua/lvim/lsp/manager.lua
@@ -125,8 +125,10 @@ function M.setup(server_name, user_config)
     end
   end
 
-  local config = resolve_config(server_name, resolve_mason_config(server_name), user_config)
-  launch_server(server_name, config)
+  if registry.is_installed(pkg_name) then
+    local config = resolve_config(server_name, resolve_mason_config(server_name), user_config)
+    launch_server(server_name, config)
+  end
 end
 
 return M
```

Thanks for your help.

**Reply by kylo252** [reply]
Sorry for the late reply, so I personally have `clangd` to be manually configured and manually installed, but this is a new workflow 😄 

The problem with this approach you're suggesting is that it's rather hard to cover some certain behaviors: how to actually launch the server when it's finished installing? see https://github.com/LunarVim/LunarVim/blob/b3bceadfddce9843850cb375edb9d470ca75b61b/lua/lvim/lsp/manager.lua#L112-L121

I think an easier alternative is to control the automatic configuration instead, e.g. [generate for certain servers and/or filetypes](https://github.com/LunarVim/LunarVim/blob/b3bceadfddce9843850cb375edb9d470ca75b61b/lua/lvim/lsp/templates.lua#L55-L74), or maybe skip the templates altogether. When you want to use LSP for a new filetype, you just need to make sure to call (it will also handle the async install for you and launch the server when ready)
```lua 
require("lvim.lsp.manager").setup("my_server", override_opts)
```

**Reply by zenofile** [reply]
Automatically installing and launching servers worked just fine in my short testing with the patch applied, so were manually installed ones.

I did, however, not test any externally installed servers, but I assumed they work just fine when set up manually without going through `lvim.lsp.manager`.

Your solution is probably better, which I will keep in mind when something breaks. This was just meant as a quick workaround.
